packages/pg-pool/index.js

```diff
@@ -116,11 +116,22 @@ class Pool extends EventEmitter {
     const pendingItem = this._pendingQueue.shift()
     if (this._idle.length) {
       const idleItem = this._idle.pop()
-      clearTimeout(idleItem.timeoutId)
       const client = idleItem.client
-      const idleListener = idleItem.idleListener
+      clearTimeout(idleItem.timeoutId)
+
+      // TODO(bmc): we need a better state represetation on the client itself
+      // to indicate if it's connecting, idle, running a query, closed, etc...
+      // as right now this hacky fix will only work for the JS client
+      // since the native client doesn't _have_ a connection property
 
-      return this._acquireClient(client, pendingItem, idleListener, false)
+      // remove this client, it's died in the background - this can happen
+      // in aws lambdas - they go idle and the streams are closed without an error
+      if (client.connection && client.connection.stream.readyState !== 'open') {
```

Comment:
I'm seeing some mention online that `stream.readyState` is deprecated/unsupported.  It might be safer to do this instead:

```
(client.connection.stream.readable && client.connection.stream.writable)
```

When experimenting w/ things locally, I'm finding that these properties change as soon as you call `stream.end()`.

Reply:
yeah those properties don't exist in all supported versions of node that pg supports.  stream.readyState _does_ exist in all supported versions.

Reply:
> When experimenting w/ things locally, I'm finding that these properties change as soon as you call `stream.end()`.

you should check if `readyState` changes as well (I think it does) - reason is any official patch is gonna need to work in node 8, 10, 12, and 13.  So if its "officially deprecated" in node but still works, I'd use it (as it probably _wasn't_ deprecated in older versions of node).  Typically I do that, write a bunch of tests for the feature, and then if a test breaks on an edge release of node I either decide to drop old version support at that point or make a compatibilty branch in the code that will handle both old and new versions of node.

Reply:
Sure -- I'll give that a try after the initial experiment.  Speaking of which: you can find the experiment within the `knex` pooling here:

https://github.com/briandamaged/knex/tree/experiment/aws-disconnections

And, here's the link to the specific commit:

https://github.com/briandamaged/knex/commit/1490611ad9fca11e42ccfeb5fdfe4e482ce6bcbc

So, I think this roughly aligns w/ your implementation.  (Aside from the `readyState` piece)

Reply:
yah totally actually I don't wanna burden you w/ backwards compatiblity - if you find a way to make it work in lambda w/ a version of node, i'll do my best to back port it and "production-ify" it for this library.  Mostly it's just hard for me to isolate & test since I don't have a lambda nor much experience w/ them so me setting all that testing infrastrucutre up is a huge time sync.  But if you figure out the root cause, I'll help any way I can getting a fix out

Reply:
Ah -- yeah, I currently don't have a way to recreate the issue, either.  I just know that Knex's current implementation relies upon Events to detect the closed connections.  So, this just seems like a possible way to work around that issue.

Hopefully we'll hear some good news from this experiment!  If so, then we can figure out how to formulate the concepts into something that is production-worthy.  (One thought: we might even need to swap out implementations depending upon Node version.  Ex: use `readable` / `writable` if the Node version supports those properties; otherwise, fall back to `readyState`)